src/main/java/forceitembattle/listener/PortalListener.java

```diff
@@ -0,0 +1,124 @@
+            }
+        }
+    }
+
+    private void teleportPlayerRandomly(Player player) {
+        Location existingLocation = this.findExistingLocation(player);
+        if (existingLocation != null) {
+            player.teleport(existingLocation);
+            return;
+        }
+
+        World world = player.getWorld();
+
+        int xOffset = random.nextBoolean() ? random.nextInt(5001) + 5000 : -(random.nextInt(5001) + 5000);
+        int zOffset = random.nextBoolean() ? random.nextInt(5001) + 5000 : -(random.nextInt(5001) + 5000);
+
+        Location currentLocation = player.getLocation();
+        Location newLocation = new Location(world, currentLocation.getX() + xOffset, currentLocation.getY(), currentLocation.getZ() + zOffset);
+        newLocation.setY(world.getHighestBlockYAt(newLocation) + 1);
+
+        Location blockLocation = newLocation.clone().subtract(0, 1, 0);
+        Block block = blockLocation.getBlock();
+        if (!block.getType().isBlock()) {
+            block.setType(Material.STONE);
+        }
+
+        playerTeleporterLocations.get(player.getUniqueId()).add(new TeleporterLocation(currentLocation, newLocation));
+        player.teleport(newLocation);
+    }
+
+    @Nullable
+    private Location findExistingLocation(Player player) {
+        Location playerLocation = player.getLocation();
+
+        for (TeleporterLocation teleporterLocation : playerTeleporterLocations.computeIfAbsent(player.getUniqueId(), k -> new ArrayList<>())) {
+            if (teleporterLocation.isClose(playerLocation)) {
+                return teleporterLocation.portalLocation;
+            }
+        }
+        return null;
+    }
+
+    private record TeleporterLocation (Location portalLocation, Location destinationLocation) {
+
+        public boolean isClose(Location location) {
+            return portalLocation.distanceSquared(location) <= 625;
+        }
+    }
+
+
+    @EventHandler
+    public void onChangedWorld(PlayerChangedWorldEvent event) {
+        if (!this.plugin.getGamemanager().isMidGame()) {
+            return;
+        }
+        Player player = event.getPlayer();
+        if (this.playerEndLocations.containsKey(player.getUniqueId())) {
+            player.teleport(this.playerEndLocations.get(player.getUniqueId()));
+            return;
+        }
+
+        if (player.getWorld().getName().equals("world_the_end")) {
+            Location spawnLocation = player.getLocation();
+
+            int xOffset = random.nextBoolean() ? random.nextInt(10_001) + 5000 : -(random.nextInt(10_001) + 5000);
+            int zOffset = random.nextBoolean() ? random.nextInt(10_001) + 5000 : -(random.nextInt(10_001) + 5000);
+
+            Location currentLocation = player.getLocation();
+            Location newLocation = new Location(player.getWorld(), currentLocation.getX() + xOffset, currentLocation.getY(), currentLocation.getZ() + zOffset);
+            newLocation.setY(player.getWorld().getHighestBlockYAt(newLocation) + 1);
```

Comment:
What if the random chosen location is above the void?

1. I'm not sure what this method returns then.
2. You seem to be missing to place a block under the users feet

Reply:
We have custom end generation without void ever, so that is not needed